<a href="https://colab.research.google.com/github/Javier-Gonzalez-Valle/Javier-Gonzalez-Valle/blob/main/CasoPracticoAnalizandoData%20Set%20P%C3%BAblico_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://marketing4ecommerce.net/wp-content/uploads/2015/09/logo-iebs.jpg" style="float:right" width="400">

# Introducción a los lenguajes de programación

## Caso Práctico: Estudio de un dataset

### Javier González Valle



In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Dataset

En esta libreta vamos a trabajar con un conjunto de datos obtenido de [Kaggle](https://www.kaggle.com/abecklas/fifa-world-cup?select=WorldCupMatches.csv) que contiene los datos de la FIFA de la copa del mundo. Recomendamos acceder a este enlace para descargar los ficheros a trabajar, aunque también disponéis de los archivos ya descargados en la sección de recursos. Utilizaremos estos tres ficheros:

- WorldCupMatches.csv: Datos de los partidos disputados
- WorldCupPlayers.csv: Datos de los jugadores que han jugado en cada partido
- WorldCups.csv: Datos de las copas disputadas y los resultados

# Introducción

Mediante esta libreta vamos a explorar los datos almacenados en estos tres ficheros y a responder preguntas haciendo operaciones y transformaciones con pandas. Además de completar las celdas de código para responder a los ejercicios, cuando así se indique se deberá añadir una celda de tipo _markdown_ explicando con texto enriquecido la información o las conclusiones extraídas.

In [ ]:
import pandas as pd

In [ ]:
# Cargamos el conjunto de datos WorldCupMatches
df_partidos = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Clase 2/WorldCupMatches.csv")


**NOTA:** en una actividad anterior completamos los ejercicios del 1 al 3, copiar y pegar dicho trabajo en esta libreta con el fin de tener un report completo que analize los datos del problema. Es decir, **no es necesario volver a realizar estos 3 primeros ejercicios**.

## 1. Exploración

Usar las funciones `head`, `info` y `describe` para explorar el dataframe `df_partidos`. **Crear una celda de tipo markdown** y comentar brevemente la información almacenada en cada dataframe (columnas y tipo de datos).

_No es necesario hacer una documentación exhaustiva, tan solo comentar aspectos principales como qué tipo de información almacena el DataFrame, cuántas filas y columnas tiene, y comentar las columnas que se consideren más relevantes._

In [ ]:
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

Con la función head() podemos observar las cinco primeras filas del dataframe.

In [ ]:
df_partidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2          

In [ ]:
df_partidos.dtypes

Year                    float64
Datetime                 object
Stage                    object
Stadium                  object
City                     object
Home Team Name           object
Home Team Goals         float64
Away Team Goals         float64
Away Team Name           object
Win conditions           object
Attendance              float64
Half-time Home Goals    float64
Half-time Away Goals    float64
Referee                  object
Assistant 1              object
Assistant 2              object
RoundID                 float64
MatchID                 float64
Home Team Initials       object
Away Team Initials       object
dtype: object

Con la función info(), como observamos nos da listada las columnas existentes, tenemos un total de 20 columnas, y además nos prpoporciona que Dtype construye los datos de esas columnas, es decir, que tipo de dato tenemos. En este caso, tenemos float64, para números decimales, y object para datos categóricos.

In [ ]:
df_partidos.describe()

Year  Home Team Goals  Away Team Goals     Attendance  \
count   852.000000       852.000000       852.000000     850.000000   
mean   1985.089202         1.811033         1.022300   45164.800000   
std      22.448825         1.610255         1.087573   23485.249247   
min    1930.000000         0.000000         0.000000    2000.000000   
25%    1970.000000         1.000000         0.000000   30000.000000   
50%    1990.000000         2.000000         1.000000   41579.500000   
75%    2002.000000         3.000000         2.000000   61374.500000   
max    2014.000000        10.000000         7.000000  173850.000000   

       Half-time Home Goals  Half-time Away Goals       RoundID       MatchID  
count            852.000000            852.000000  8.520000e+02  8.520000e+02  
mean               0.708920              0.428404  1.066177e+07  6.134687e+07  
std                0.937414              0.691252  2.729613e+07  1.110572e+08  
min                0.000000              0.000000  2.010000e+02  2.500000e+01  
25%                0.000000              0.000000  2.620000e+02  1.188750e+03  
50%                0.000000              0.000000  3.370000e+02  2.191000e+03  
75%                1.000000              1.000000  2.497220e+05  4.395006e+07  
max                6.000000              5.000000  9.741060e+07  3.001865e+08

LA función describe() nos da un resumen descriptivo de las variables numéricas. El número de datos completos, la media, la Desviación típica, los tres cuartiles para dividir nuestra muestra en cuatro trozos, y el máximo y mínimo.

Para observar recibir los resultados de los objetos iteramos de esta forma:

In [ ]:
df_partidos.describe(include='object')

Datetime        Stage         Stadium          City  \
count                    852          852             852           852   
unique                   602           23             181           151   
top     27 May 1934 - 16:30   Round of 16  Estadio Azteca  Mexico City    
freq                       8           72              19            23   

       Home Team Name Away Team Name Win conditions                Referee  \
count             852            852            852                    852   
unique             78             83             43                    366   
top            Brazil         Mexico                 Ravshan IRMATOV (UZB)   
freq               82             38            787                     10   

                 Assistant 1               Assistant 2 Home Team Initials  \
count                    852                       852                852   
unique                   387                       408                 77   
top     ACHIK Redouane (MAR)  KOCHKAROV Bakhadyr (KGZ)                BRA   
freq                       7                        10                 82   

       Away Team Initials  
count                 852  
unique                 82  
top                   MEX  
freq                   38

Para observar la descripción de los objetos necesitamos utilizar el código previo.

In [ ]:
df_partidos.shape

(4572, 20)

In [ ]:
df_partidos.size

91440

Con shape y size pues obtenemos información más detallada del df. En este caso sabemos que tiene 4572 filas y 20 columnas y un total de datos de 91440.

## 2. Limpieza de datos

Una de las cosas que llama la atención es la alta presencia de valores perdidos en el DataFrame `df_partidos`. Cuando trabajamos con datos obtenidos del mundo real siempre nos toparemos con problemas relacionados con la medición, captura o almacenamiento de dicha información.

Localiza las filas con valores perdido. Analizar y **documentar en una nueva celda de tipo markdown** a qué se deben estos valores perdidos. Finalmente usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) sobre el DataFrame `df_partidos` para eliminar los valores perdidos del DataFrame, y almacena el resultado en `df_partidos` de nuevo.

In [ ]:
df_partidos.isnull()

Year  Datetime  Stage  Stadium   City  Home Team Name  Home Team Goals  \
0     False     False  False    False  False           False            False   
1     False     False  False    False  False           False            False   
2     False     False  False    False  False           False            False   
3     False     False  False    False  False           False            False   
4     False     False  False    False  False           False            False   
...     ...       ...    ...      ...    ...             ...              ...   
4567   True      True   True     True   True            True             True   
4568   True      True   True     True   True            True             True   
4569   True      True   True     True   True            True             True   
4570   True      True   True     True   True            True             True   
4571   True      True   True     True   True            True             True   

      Away Team Goals  Away Team Name  Win conditions  Attendance  \
0               False           False           False       False   
1               False           False           False       False   
2               False           False           False       False   
3               False           False           False       False   
4               False           False           False       False   
...               ...             ...             ...         ...   
4567             True            True            True        True   
4568             True            True            True        True   
4569             True            True            True        True   
4570             True            True            True        True   
4571             True            True            True        True   

      Half-time Home Goals  Half-time Away Goals  Referee  Assistant 1  \
0                    False                 False    False        False   
1                    False                 False    False        False   
2                    False                 False    False        False   
3                    False                 False    False        False   
4                    False                 False    False        False   
...                    ...                   ...      ...          ...   
4567                  True                  True     True         True   
4568                  True                  True     True         True   
4569                  True                  True     True         True   
4570                  True                  True     True         True   
4571                  True                  True     True         True   

      Assistant 2  RoundID  MatchID  Home Team Initials  Away Team Initials  
0           False    False    False               False               False  
1           False    False    False               False               False  
2           False    False    False               False               False  
3           False    False    False               False               False  
4           False    False    False               False               False  
...           ...      ...      ...                 ...                 ...  
4567         True     True     True                True                True  
4568         True     True     True                True                True  
4569         True     True     True                True                True  
4570         True     True     True                True                True  
4571         True     True     True                True                True  

[4572 rows x 20 columns]

In [ ]:
df_partidos.isnull().sum()

Year                    3720
Datetime                3720
Stage                   3720
Stadium                 3720
City                    3720
Home Team Name          3720
Home Team Goals         3720
Away Team Goals         3720
Away Team Name          3720
Win conditions          3720
Attendance              3722
Half-time Home Goals    3720
Half-time Away Goals    3720
Referee                 3720
Assistant 1             3720
Assistant 2             3720
RoundID                 3720
MatchID                 3720
Home Team Initials      3720
Away Team Initials      3720
dtype: int64

En base a este código, observamos que hay 3720 valores null en todas las columnas.
Parece ser, que estos valores se concentran al final del df.Por lo que aparentemente parece que a la hora de formar este dataframe se seleccionaron más filas de las que estaban completas. Para eliminarlas y limpiar el df realizaremos lo siguiente.

In [ ]:
df_partidos = df_partidos.dropna()
df_partidos

Year              Datetime                     Stage  \
0    1930.0  13 Jul 1930 - 15:00                    Group 1   
1    1930.0  13 Jul 1930 - 15:00                    Group 4   
2    1930.0  14 Jul 1930 - 12:45                    Group 2   
3    1930.0  14 Jul 1930 - 14:50                    Group 3   
4    1930.0  15 Jul 1930 - 16:00                    Group 1   
..      ...                   ...                       ...   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
0                Pocitos      Montevideo          France              4.0   
1         Parque Central      Montevideo             USA              3.0   
2         Parque Central      Montevideo      Yugoslavia              2.0   
3                Pocitos      Montevideo         Romania              3.0   
4         Parque Central      Montevideo       Argentina              1.0   
..                   ...              ...            ...              ...   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
0                1.0         Mexico                                          
1                0.0        Belgium                                          
2                1.0         Brazil                                          
3                1.0           Peru                                          
4                0.0         France                                          
..               ...            ...                                    ...   
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
0        4444.0                   3.0                   0.0   
1       18346.0                   2.0                   0.0   
2       24059.0                   2.0                   0.0   
3        2549.0                   1.0                   0.0   
4       23409.0                   0.0                   0.0   
..          ...                   ...                   ...   
847     51179.0                   0.0                   0.0   
848     58141.0                   0.0                   5.0   
849     63267.0                   0.0                   0.0   
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   

                    Referee                  Assistant 1  \
0    LOMBARDI Domingo (URU)        CRISTOPHE Henry (BEL)   
1         MACIAS Jose (ARG)     MATEUCCI Francisco (URU)   
2       TEJADA Anibal (URU)      VALLARINO Ricardo (URU)   
3     WARNKEN Alberto (CHI)          LANGENUS Jean (BEL)   
4       REGO Gilberto (BRA)         SAUCEDO Ulises (BOL)   
..                      ...                          ...   
847   Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848   RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849      C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850   HAIMOUDI Djamel (ALG)    

## 3. Cargar datos

Al igual que hemos cargado los datos de los partidos en `df_partidos`, crear las variables `df_jugadores` y `df_copas` que contengan los dataframes correspondientes a la lectura de los csv `WorldCupPlayers` y `WorldCups` (usar `pd.read_csv)`.


In [ ]:
df_jugadores = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Clase 2/WorldCupPlayers.csv")
df_jugadores.head()

RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
3      201     1096           MEX     LUQUE Juan (MEX)       S             0   
4      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'  
3      Juan CARRENO      NaN  G70'  
4   Ernest LIBERATI      NaN   NaN

In [ ]:
df_copas = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Clase 2/WorldCups.csv")
df_copas.head()

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607

## 4. Rango temporal

El dataframe `df_copas` contiene datos de todos los mundiales disputados. ¿Cuál es el año del mundial más antiguo disputado? ¿Y el año del mundial más reciente?

**Pista:** recuerda que podemos usar funciones de agregación (`count` para contar el número de casos, `mean` para calcular el valor medio, etc.) directamente sobre columnas o Series de pandas. Por ejemplo, el siguiente código nos muestra el mayor número de goles marcado en un mundial.

```python
df_copas["GoalsScored"].max()
```

**Opcional:** Una vez localizados los años de interés (más antiguo y más reciente) visualizar las filas completas correspondientes a cada año usando la función `.loc`.

In [ ]:
mask1 = df_copas["Year"].min()
mask1

1930

In [ ]:
mask2= df_copas["Year"].max()
mask2

2014

In [ ]:
df_copas.set_index("Year").loc[1930]

Country              Uruguay
Winner               Uruguay
Runners-Up         Argentina
Third                    USA
Fourth            Yugoslavia
GoalsScored               70
QualifiedTeams            13
MatchesPlayed             18
Attendance           590.549
Name: 1930, dtype: object

In [ ]:
df_copas.set_index("Year").loc[2014]

Country                Brazil
Winner                Germany
Runners-Up          Argentina
Third             Netherlands
Fourth                 Brazil
GoalsScored               171
QualifiedTeams             32
MatchesPlayed              64
Attendance          3.386.810
Name: 2014, dtype: object

---
# Ejercicios a realizar

Los ejercicios a continuación amplian el reporte comenzado en una actividad semanal anterior. Algunos de ellos son más complejos y son opcionales, pero se recomienda que al menos se intente resolverlos pues puntuará positivamente.

---

## 5. Asistencia de público

La columna `Attendance` indica la asistencia de público en cada mundial. Pandas ha cargado esta columna como tipo string, ya que los valores indicados contienen puntos como separador de millares. En primer lugar vamos a transformar esta columna para que no contenga el caracter `.`. Posteriormente transformaremos esta columna a formato numérico usando la función [pd.to_numeric](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html).

**Nota**: la función `pd.read_csv` también permite hacer este proceso de una forma mucho más transparente, usando el argumento `thousands`. Ver la documentación [aquí](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [ ]:
df_copas["Attendance"] = pd.to_numeric(df_copas["Attendance"].str.replace(".", ""))
# df_copas = pd.read_csv("WorldCups.csv", thousands=".")

<ipython-input-56-7b9d992eac5a>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_copas["Attendance"] = pd.to_numeric(df_copas["Attendance"].str.replace(".", ""))


Teniendo en cuenta todos los datos de `df_copas`, ¿Cuál fué la mayor asistencia de un mundial? ¿Cuál es la asistencia media?

In [ ]:
df_copas.head()

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed  Attendance  
0           70              13             18      590549  
1           70              16             17      363000  
2           84              15             18      375700  
3           88              13             22     1045246  
4          140              16             26      768607

In [ ]:
df_copas.agg(["max","min"])

Year    Country     Winner Runners-Up    Third      Fourth  GoalsScored  \
max  2014    Uruguay    Uruguay     Sweden      USA  Yugoslavia          171   
min  1930  Argentina  Argentina  Argentina  Austria     Austria           70   

     QualifiedTeams  MatchesPlayed  Attendance  
max              32             64     3587538  
min              13             17      363000

In [ ]:
df_copas['Attendance'].max()

3587538

In [ ]:
df_copas['Attendance'].mean()

1872882.35

## 6. Asistencia de público en función del país

Sabemos que, dependiendo del país donde se celebre la copa del mundo, ésta tiene un mayor o menor impacto o relevancia. Vamos a calcular la media de asistencia agrupando por país. Finalmente, ordenar los resultados por orden decreciente de este número medio de asistentes.

**Pista 1:** Como se muestra en el siguiente ejemplo, recuerda que podemos combinar la función [groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) con otras funciones de agregación como `mean`.

**Pista 2:** Recuerda que podemos usar la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) para ordenar un DataFrame por una de sus columnas. Por ejemplo, el siguiente código muestra los mundiales ordenados de forma decreciente por el número de goles marcados:

```python
df_copas.sort_values("GoalsScored", ascending=False)
```

In [ ]:
Attend = df_copas.groupby("Country").mean().Attendance
Attend

<ipython-input-67-2f1dabe2f227>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Attend = df_copas.groupby("Country").mean().Attendance


Country
Argentina       1545791.0
Brazil          2216028.0
Chile            893172.0
England         1563135.0
France          1580400.0
Germany         2612596.0
Italy           1439607.5
Korea/Japan     2705197.0
Mexico          1999003.0
South Africa    3178856.0
Spain           2109723.0
Sweden           819810.0
Switzerland      768607.0
USA             3587538.0
Uruguay          590549.0
Name: Attendance, dtype: float64

In [ ]:
Attend.sort_values(ascending= False)


Country
USA             3587538.0
South Africa    3178856.0
Korea/Japan     2705197.0
Germany         2612596.0
Brazil          2216028.0
Spain           2109723.0
Mexico          1999003.0
France          1580400.0
England         1563135.0
Argentina       1545791.0
Italy           1439607.5
Chile            893172.0
Sweden           819810.0
Switzerland      768607.0
Uruguay          590549.0
Name: Attendance, dtype: float64

In [ ]:
df_personas = pd.DataFrame(
    [
        ("Juan", "male", 56),
        ("Laura", "female", 23),
        ("José", "male", 46),
        ("Rosa", "female", 26)
    ], columns=["name", "gender", "age"])

# edad media de todas las personas
edad_media = df_personas["age"].mean()
# edad media agrupando por género
edad_media_por_genero = df_personas.groupby("gender").mean()[["age"]]

print(f"La edad media de las {len(df_personas)} es de {edad_media} años.")
print(f"La edad media agrupada por género es:")
edad_media_por_genero

La edad media de las 4 es de 37.75 años.
La edad media agrupada por género es:


age
gender      
female  24.5
male    51.0

In [ ]:
# Completar

## 7. País con más victorias

Vamos a analizar el número de veces que un país ha ganado la copa del mundo. Transformar el DataFrame `df_copas` para que muestre los diferentes países que han ganado al menos una copa del mundo, y ordena el DataFrame por dicho número de forma descendente usando la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

¿Ves algo extraño o digno de comentar? Pista: analiza Alemania. **Hazlo en una nueva celda de tipo markdown.**

In [ ]:
df_copas.Winner

0        Uruguay
1          Italy
2          Italy
3        Uruguay
4     Germany FR
5         Brazil
6         Brazil
7        England
8         Brazil
9     Germany FR
10     Argentina
11         Italy
12     Argentina
13    Germany FR
14        Brazil
15        France
16        Brazil
17         Italy
18         Spain
19       Germany
Name: Winner, dtype: object

In [ ]:
df_copas["Winner"].value_counts().sort_values(ascending = False)

Brazil        5
Italy         4
Germany FR    3
Uruguay       2
Argentina     2
England       1
France        1
Spain         1
Germany       1
Name: Winner, dtype: int64

Como observamos, en la lista anterior, Germany aparece dos veces. Esto se deberá a que la última copa del mundo ganada Alemania ya era un país unificado, previamente era una República Federal y una República DEmocrática dividida por el Muro de Berlín. Es por eso que existen dos valores de Alemania en este DF.

---
# Parte opcional

Los ejercicios anteriores son obligatorios y puntuan hasta un máximo de 80 puntos sobre 100.

Los siguientes ejercicios trabajan con los dataframes cargados para extraer información concreta a partir de los datos.
Estos ejercicios son opcionales, pero es recomendable intentar resolverlos para adquirir habilidades.

---

## Creación de un nuevo DataFrame

A continuación vamos a trabajar con el DataFrame `df_partidos`. Vamos a centrarnos en la información a nivel de selección, siendo indiferente si el equipo juega en casa o fuera. En este sentido, la información disponible no es la mejor para trabajar cómodamente. Por ello, es habitual realizar transformaciones a los datos en función de las preguntas que le queramos hacer. A continuación crearemos un nuevo DataFrame llamado `df_paises`que contendrá todos los paises involucrados en la copa del mundo y el año en el que participaron. Es decir, un país que haya jugado varias copas del mundo aparecerá varias veces:


||Year|Country|
|---|---|---|
|0|1930|France|
|1|1930|USA|
|...|...|..|
|37|1938|France|
|...|...|..|

**Nota:** este problema se puede afrontar como mínimo de dos maneras diferentes

1. Generando dos dataframes, uno donde la columna `Country` se corresponde con `Home Team Name`, y otro donde la columna `Country` se corresponde con `Away Team Name`. Finalmente, combinar ambos DataFrames en uno solo.
2. Usar la función [melt](https://pandas.pydata.org/docs/reference/api/pandas.melt.html) que nos permite convertir N columnas en un par de columnas llamadas `variable` y `value`, donde la primera es el nombre de una de las columnas y `value` el valor que tomó dicha variable.


In [ ]:
#alternativa 1
df_paises = (
    pd.concat([
        df_partidos[["Year", "Home Team Name"]].rename(columns={"Home Team Name": "Country"}),
        df_partidos[["Year", "Away Team Name"]].rename(columns={"Away Team Name": "Country"})
    ])
)
#alternativa 2
df_paises = (
    df_partidos
    .melt(id_vars=["Year"], value_vars=["Home Team Name", "Away Team Name"])
    .rename(columns={"value": "Country"})
    [["Year", "Country"]]
)

In [ ]:
df_paises = (
    pd.concat([
        df_partidos[["Year", "Home Team Name"]].rename(columns={"Home Team Name": "Country"}),
        df_partidos[["Year", "Away Team Name"]].rename(columns={"Away Team Name": "Country"})
    ])
)
df_paises

Year     Country
0     1930.0      France
1     1930.0         USA
2     1930.0  Yugoslavia
3     1930.0     Romania
4     1930.0   Argentina
...      ...         ...
4567     NaN         NaN
4568     NaN         NaN
4569     NaN         NaN
4570     NaN         NaN
4571     NaN         NaN

[9144 rows x 2 columns]

In [ ]:
df_paises = df_paises.dropna()
df_paises

Year      Country
0    1930.0       France
1    1930.0          USA
2    1930.0   Yugoslavia
3    1930.0      Romania
4    1930.0    Argentina
..      ...          ...
847  2014.0   Costa Rica
848  2014.0      Germany
849  2014.0    Argentina
850  2014.0  Netherlands
851  2014.0    Argentina

[1704 rows x 2 columns]

## 8. Número de países

Vamos a descubrir cuántos países han participado, al menos una vez, en una copa del mundo. Para ello vamos a trabajar con el DataFrame construido anteriormente y a usar la función de agregación [nunique](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nunique.html).

In [ ]:
df_paises.nunique()

Year       20
Country    83
dtype: int64

Con esta función observamos que hay 83 países que han participado al menos 1 vez en una copa del mundo. También nos da que ha habido 20 copas del mundo.

## 9. Número de partidos por país

Ahora vamos a descubrir cuántos partidos ha jugado cada país. Operar con el DataFrame para que muestre un país por fila y una columna que indique el número de partidos que ha jugado. Ordenar el DataFrame por dicha columna de forma descendente.

In [ ]:
df_paises["Country"].value_counts().sort_values(ascending = False)

Brazil                        108
Italy                          83
Argentina                      81
Germany FR                     62
England                        62
                             ... 
Canada                          3
rn">Bosnia and Herzegovina      3
Serbia                          3
Israel                          3
Dutch East Indies               1
Name: Country, Length: 83, dtype: int64

## 10. Partidos jugados por España

Vamos a trabajar a continuación con el DataFrame `df_partidos`. Seleccionar aquellos partidos (filas del dataframe) en los que España ha jugado como `Home Team`, es decir, donde la columna `Home Team` tiene el valor `España`. Recuerda que para la operación de filtrado hay que usar la función [.loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html).


A continuación, haz una nueva selección en los que España ha jugado como `Away Team`, es decir, donde la columna `Away Team` tiene el valor `España`.

**Opcional:** ¿Podrías hacer una sola selección donde España haya jugado como `Home Team` o como `Away Team`?. Pista: consulta la documentación del operador de [disyunción](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing).

In [ ]:
df_partidos = df_partidos.dropna()
df_partidos.sample(2)

Year              Datetime           Stage          Stadium  \
283  1978.0  07 Jun 1978 - 13:45          Group 3  Jose Amalfitani   
406  1986.0  22 Jun 1986 - 16:00   Quarter-finals       Cuauhtemoc   

              City Home Team Name  Home Team Goals  Away Team Goals  \
283  Buenos Aires         Austria              1.0              0.0   
406        Puebla           Spain              1.0              1.0   

    Away Team Name                     Win conditions  Attendance  \
283         Sweden                                        41424.0   
406        Belgium  Belgium win on penalties (4 - 5)      45000.0   

     Half-time Home Goals  Half-time Away Goals                   Referee  \
283                   1.0                   0.0      CORVER Charles (NED)   
406                   0.0                   0.0  KIRSCHEN Siegfried (GER)   

                      Assistant 1            Assistant 2  RoundID  MatchID  \
283        MAKSIMOVIC Dusan (SCG)      SEOUDI Hedi (TUN)    278.0   2224.0   
406  CODESAL MENDEZ Edgardo (MEX)  BRUMMEIER Horst (AUT)    714.0    421.0   

    Home Team Initials Away Team Initials  
283                AUT                SWE  
406                ESP                BEL

In [ ]:
df_partidos.set_index("Home Team Name").loc["Spain"]

Year               Datetime              Stage  \
Home Team Name                                                     
Spain           1934.0   27 May 1934 - 16:30   Preliminary round   
Spain           1950.0   25 Jun 1950 - 15:00             Group 2   
Spain           1950.0   29 Jun 1950 - 15:00             Group 2   
Spain           1950.0   02 Jul 1950 - 15:00             Group 2   
Spain           1962.0   03 Jun 1962 - 15:00             Group 3   
Spain           1966.0   15 Jul 1966 - 19:30             Group 2   
Spain           1978.0   11 Jun 1978 - 13:45             Group 3   
Spain           1982.0   16 Jun 1982 - 21:00             Group 5   
Spain           1982.0   20 Jun 1982 - 21:00             Group 5   
Spain           1982.0   05 Jul 1982 - 21:00             Group 2   
Spain           1986.0   01 Jun 1986 - 12:00             Group D   
Spain           1986.0   22 Jun 1986 - 16:00      Quarter-finals   
Spain           1990.0   26 Jun 1990 - 17:00         Round of 16   
Spain           1994.0   17 Jun 1994 - 19:30             Group C   
Spain           1994.0   02 Jul 1994 - 16:30         Round of 16   
Spain           1998.0   13 Jun 1998 - 14:30             Group D   
Spain           1998.0   19 Jun 1998 - 21:00             Group D   
Spain           1998.0   24 Jun 1998 - 21:00             Group D   
Spain           2002.0   02 Jun 2002 - 20:30             Group B   
Spain           2002.0  07 June 2002 - 18:00             Group B   
Spain           2002.0   16 Jun 2002 - 20:30         Round of 16   
Spain           2002.0   22 Jun 2002 - 15:30      Quarter-finals   
Spain           2006.0   14 Jun 2006 - 15:00             Group H   
Spain           2006.0   19 Jun 2006 - 21:00             Group H   
Spain           2006.0   27 Jun 2006 - 21:00         Round of 16   
Spain           2010.0   16 Jun 2010 - 16:00             Group H   
Spain           2010.0   21 Jun 2010 - 20:30             Group H   
Spain           2010.0   29 Jun 2010 - 20:30         Round of 16   
Spain           2014.0   13 Jun 2014 - 16:00             Group B   
Spain           2014.0   18 Jun 2014 - 16:00             Group B   

                                                  Stadium             City  \
Home Team Name                                                               
Spain                                      Luigi Ferraris           Genoa    
Spain                                    Durival de Brito        Curitiba    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain                                   Estadio Sausalito    Vina Del Mar    
Spain                                        Hillsborough       Sheffield    
Spain                                     Jose Amalfitani    Buenos Aires    
Spain                                       Luis Casanova        Valencia    
Spain                                       Luis Casanova        Valencia    
Spain                                   Santiago Bernabeu          Madrid    
Spain                                             Jalisco     Guadalajara    
Spain                                          Cuauhtemoc          Puebla    
Spain                              Marc Antonio Bentegodi          Verona    
Spain                                         Cotton Bowl          Dallas    
Spain                                         RFK Stadium   Washington Dc    
Spain                                        La Beaujoire          Nantes    
Spain                             Stade Geoffroy Guichard   Saint-Etienne    
Spain                                Stade Felix Bollaert            Lens    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                            Jeonju World Cup Stadium          Jeonju    
Spain                             Suwon World Cup Stadium           Suwon    
Spain                           Gwangju World C

In [ ]:
df_partidos.set_index("Away Team Name").loc["Spain"]

Year              Datetime           Stage  \
Away Team Name                                                 
Spain           1934.0  31 May 1934 - 16:30   Quarter-finals   
Spain           1934.0  01 Jun 1934 - 16:30   Quarter-finals   
Spain           1950.0  09 Jul 1950 - 15:00          Group 6   
Spain           1950.0  13 Jul 1950 - 15:00          Group 6   
Spain           1950.0  16 Jul 1950 - 15:00          Group 6   
Spain           1962.0  31 May 1962 - 15:00          Group 3   
Spain           1962.0  06 Jun 1962 - 15:00          Group 3   
Spain           1966.0  13 Jul 1966 - 19:30          Group 2   
Spain           1966.0  20 Jul 1966 - 19:30          Group 2   
Spain           1978.0  03 Jun 1978 - 13:45          Group 3   
Spain           1978.0  07 Jun 1978 - 13:45          Group 3   
Spain           1982.0  25 Jun 1982 - 21:00          Group 5   
Spain           1982.0  02 Jul 1982 - 21:00          Group 2   
Spain           1986.0  07 Jun 1986 - 12:00          Group D   
Spain           1986.0  12 Jun 1986 - 12:00          Group D   
Spain           1986.0  18 Jun 1986 - 16:00      Round of 16   
Spain           1990.0  13 Jun 1990 - 17:00          Group E   
Spain           1990.0  17 Jun 1990 - 21:00          Group E   
Spain           1990.0  21 Jun 1990 - 17:00          Group E   
Spain           1994.0  21 Jun 1994 - 16:00          Group C   
Spain           1994.0  27 Jun 1994 - 16:00          Group C   
Spain           1994.0  09 Jul 1994 - 12:00   Quarter-finals   
Spain           2002.0  12 Jun 2002 - 20:30          Group B   
Spain           2006.0  23 Jun 2006 - 16:00          Group H   
Spain           2010.0  25 Jun 2010 - 20:30          Group H   
Spain           2010.0  03 Jul 2010 - 20:30   Quarter-finals   
Spain           2010.0  07 Jul 2010 - 20:30      Semi-finals   
Spain           2010.0  11 Jul 2010 - 20:30            Final   
Spain           2014.0  23 Jun 2014 - 13:00          Group B   

                                                  Stadium               City  \
Away Team Name                                                                 
Spain                                      Giovanni Berta          Florence    
Spain                                      Giovanni Berta          Florence    
Spain                                            Pacaembu         Sao Paulo    
Spain           Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
Spain                                            Pacaembu         Sao Paulo    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                          Villa Park        Birmingham    
Spain                                          Villa Park        Birmingham    
Spain                                     Jose Amalfitani      Buenos Aires    
Spain                          Estadio Jos� Mar�a Minella     Mar Del Plata    
Spain                                       Luis Casanova          Valencia    
Spain                                   Santiago Bernabeu            Madrid    
Spain                                       Tres de Marzo       Guadalajara    
Spain                                         Tecnologico         Monterrey    
Spain                                 Estadio Corregidora         Queretaro    
Spain                                         Dacia Arena             Udine    
Spain                                         Dacia Arena             Udine    
Spain                              Marc Antonio Bentegodi            Verona    
Spain                                       Soldier Field           Chicago    
Spain                                       Soldier Field           Chicago    
Spain                                     Foxboro Stadium            Boston    
Spain                           Daejeon World Cup Stadium           Daejeon    
Spain                            

In [ ]:
df_partidos.set_index("Home Team Name","Away Team Name").loc["Spain"]

<ipython-input-110-758eaa8bfdf2>:1: FutureWarning: In a future version of pandas all arguments of DataFrame.set_index except for the argument 'keys' will be keyword-only.
  df_partidos.set_index("Home Team Name","Away Team Name").loc["Spain"]


Year               Datetime              Stage  \
Home Team Name                                                     
Spain           1934.0   27 May 1934 - 16:30   Preliminary round   
Spain           1950.0   25 Jun 1950 - 15:00             Group 2   
Spain           1950.0   29 Jun 1950 - 15:00             Group 2   
Spain           1950.0   02 Jul 1950 - 15:00             Group 2   
Spain           1962.0   03 Jun 1962 - 15:00             Group 3   
Spain           1966.0   15 Jul 1966 - 19:30             Group 2   
Spain           1978.0   11 Jun 1978 - 13:45             Group 3   
Spain           1982.0   16 Jun 1982 - 21:00             Group 5   
Spain           1982.0   20 Jun 1982 - 21:00             Group 5   
Spain           1982.0   05 Jul 1982 - 21:00             Group 2   
Spain           1986.0   01 Jun 1986 - 12:00             Group D   
Spain           1986.0   22 Jun 1986 - 16:00      Quarter-finals   
Spain           1990.0   26 Jun 1990 - 17:00         Round of 16   
Spain           1994.0   17 Jun 1994 - 19:30             Group C   
Spain           1994.0   02 Jul 1994 - 16:30         Round of 16   
Spain           1998.0   13 Jun 1998 - 14:30             Group D   
Spain           1998.0   19 Jun 1998 - 21:00             Group D   
Spain           1998.0   24 Jun 1998 - 21:00             Group D   
Spain           2002.0   02 Jun 2002 - 20:30             Group B   
Spain           2002.0  07 June 2002 - 18:00             Group B   
Spain           2002.0   16 Jun 2002 - 20:30         Round of 16   
Spain           2002.0   22 Jun 2002 - 15:30      Quarter-finals   
Spain           2006.0   14 Jun 2006 - 15:00             Group H   
Spain           2006.0   19 Jun 2006 - 21:00             Group H   
Spain           2006.0   27 Jun 2006 - 21:00         Round of 16   
Spain           2010.0   16 Jun 2010 - 16:00             Group H   
Spain           2010.0   21 Jun 2010 - 20:30             Group H   
Spain           2010.0   29 Jun 2010 - 20:30         Round of 16   
Spain           2014.0   13 Jun 2014 - 16:00             Group B   
Spain           2014.0   18 Jun 2014 - 16:00             Group B   

                                                  Stadium             City  \
Home Team Name                                                               
Spain                                      Luigi Ferraris           Genoa    
Spain                                    Durival de Brito        Curitiba    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain                                   Estadio Sausalito    Vina Del Mar    
Spain                                        Hillsborough       Sheffield    
Spain                                     Jose Amalfitani    Buenos Aires    
Spain                                       Luis Casanova        Valencia    
Spain                                       Luis Casanova        Valencia    
Spain                                   Santiago Bernabeu          Madrid    
Spain                                             Jalisco     Guadalajara    
Spain                                          Cuauhtemoc          Puebla    
Spain                              Marc Antonio Bentegodi          Verona    
Spain                                         Cotton Bowl          Dallas    
Spain                                         RFK Stadium   Washington Dc    
Spain                                        La Beaujoire          Nantes    
Spain                             Stade Geoffroy Guichard   Saint-Etienne    
Spain                                Stade Felix Bollaert            Lens    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                            Jeonju World Cup Stadium          Jeonju    
Spain                             Suwon World Cup Stadium           Suwon    
Spain                           Gwangju World C

## Calcular el vencedor de cada partido

El DataFrame `df_partidos` tiene la información de cada partido, incluyendo los goles marcados por cada equipo, pero no disponemos de una columna que indique si un equipo ganó o perdió. Vamos a crear una columna llamada `Ganador` que indique el nombre del equipo vencedor (con más goles). También podemos encontrar empates, en cuyo caso usaremos `pd.NA` para indicar que desconocemos el vencedor de dicho partido.

**Nota:** Se puede resolver realizando varias asignaciones, o una sola utilizando la función [pd.where](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.where.html).

In [ ]:
# alternativa 1
df_partidos["Ganador"] = df_partidos["Home Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA

# alternativa 2
df_partidos["Ganador"] = (
    df_partidos["Home Team Name"].where(
        df_partidos["Home Team Goals"] > df_partidos["Away Team Goals"],
        df_partidos["Away Team Name"].where(df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], pd.NA)
    )
)

In [ ]:
df_partidos["Ganador"] = df_partidos["Home Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA
df_partidos.head()

<ipython-input-112-cd990aa30540>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_partidos["Ganador"] = df_partidos["Home Team Name"]


Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Home Goals  Half-time Away Goals  \
0                 ...                   3.0                   0.0   
1                 ...                   2.0                   0.0   
2                 ...                   2.0                   0.0   
3                 ...                   1.0                   0.0   
4                 ...                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2 RoundID  MatchID  Home Team Initials  \
0         REGO Gilberto (BRA)   201.0   1096.0                 FRA   
1       WARNKEN Alberto (CHI)   201.0   1090.0                 USA   
2         BALWAY Thomas (FRA)   201.0   1093.0                 YUG   
3    MATEUCCI Francisco (URU)   201.0   1098.0                 ROU   
4  RADULESCU Constantin (ROU)   201.0   1085.0                 ARG   

  Away Team Initials     Ganador  
0                MEX      France  
1                BEL         USA  
2                BRA  Yugoslavia  
3                PER     Romania  
4                FRA   Argentina  

[5 rows x 21 columns]

## 11. Número de victorias por país

Calcular para cada país el número de partidos ganados. Para ello, usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) para eliminar las filas correspondientes a partidos donde la columna `Ganador` es un valor perdido (`na`). Ordenar los países por número de victorias de forma descendente.

In [ ]:
df_partidos = df_partidos.dropna()
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Home Goals  Half-time Away Goals  \
0                 ...                   3.0                   0.0   
1                 ...                   2.0                   0.0   
2                 ...                   2.0                   0.0   
3                 ...                   1.0                   0.0   
4                 ...                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2 RoundID  MatchID  Home Team Initials  \
0         REGO Gilberto (BRA)   201.0   1096.0                 FRA   
1       WARNKEN Alberto (CHI)   201.0   1090.0                 USA   
2         BALWAY Thomas (FRA)   201.0   1093.0                 YUG   
3    MATEUCCI Francisco (URU)   201.0   1098.0                 ROU   
4  RADULESCU Constantin (ROU)   201.0   1085.0                 ARG   

  Away Team Initials     Ganador  
0                MEX      France  
1                BEL         USA  
2                BRA  Yugoslavia  
3                PER     Romania  
4                FRA   Argentina  

[5 rows x 21 columns]

In [ ]:
df_partidos["Ganador"].value_counts().sort_values(ascending = False)

Brazil                        71
Italy                         45
Argentina                     44
Germany FR                    36
Germany                       32
                              ..
Jamaica                        1
Wales                          1
Korea DPR                      1
Tunisia                        1
rn">Bosnia and Herzegovina     1
Name: Ganador, Length: 64, dtype: int64

## Opcional 1

Calcular para cada país el ratio de victorias, calculado como el número de partidos ganados dividido por el número de partidos jugados. Ordenar los países por ratio de forma descendente.

**Pista:** Si obtenemos dos dataframes, uno con las victorias por cada país y otro con el número total de partidos jugados, podemos utilizar la función [pd.merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) para combinar ambos en un solo DataFrame, y despues crear una nueva columna que sea el ratio.

In [ ]:
df_partidos["MatchID"].value_counts()

300186497.0    2
300186502.0    2
300186491.0    2
300186462.0    2
300186485.0    2
              ..
2170.0         1
2067.0         1
2065.0         1
2064.0         1
564.0          1
Name: MatchID, Length: 649, dtype: int64

In [ ]:
#número de partidos por país
total_partidos = df_paises["Country"].value_counts().sort_values(ascending = False)
print(total_partidos)



Brazil                        108
Italy                          83
Argentina                      81
Germany FR                     62
England                        62
                             ... 
Canada                          3
rn">Bosnia and Herzegovina      3
Serbia                          3
Israel                          3
Dutch East Indies               1
Name: Country, Length: 83, dtype: int64


In [ ]:
#número de partidos ganados
total_ganados = df_partidos["Ganador"].value_counts().sort_values(ascending = False)
print(total_ganados)

Brazil                        71
Italy                         45
Argentina                     44
Germany FR                    36
Germany                       32
                              ..
Jamaica                        1
Wales                          1
Korea DPR                      1
Tunisia                        1
rn">Bosnia and Herzegovina     1
Name: Ganador, Length: 64, dtype: int64


In [ ]:
df_partidos["Ratio"] = df_partidos["Home Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA

In [ ]:
ratio = (total_partidos)/(total_ganados)
print(ratio)
#ratio.sort_values(ascending = False)
#print(ratio)

Algeria                       4.666667
Angola                             NaN
Argentina                     1.840909
Australia                     6.500000
Austria                       2.416667
                                ...   
rn">Bosnia and Herzegovina    3.000000
rn">Republic of Ireland       6.500000
rn">Serbia and Montenegro          NaN
rn">Trinidad and Tobago            NaN
rn">United Arab Emirates           NaN
Length: 83, dtype: float64


## Opcional 2

¿Se te ocurre alguna pregunta adicional que hacerle a los datos? Utiliza una celda de tipo markdown para documentarla y otra de código para implementarla y ver así la respuesta.